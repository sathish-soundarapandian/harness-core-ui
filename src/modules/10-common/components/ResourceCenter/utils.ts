/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import moment from 'moment'
import type { UserInfo } from 'services/cd-ng'

export const ON_PREM_RELEASE_NODE_LINK = 'https://ngdocs.harness.io/article/556wy85kbo-harness-on-prem-release-notes'
export const SAAS_RELEASE_NODE_LINK = 'https://ngdocs.harness.io/article/7zkchy5lhj-harness-saa-s-release-notes-2022'

export const HARNESS_SEARCH_LINK = 'https://harness.io/search/'
export const HARNESS_UNIVERISITY_LINK = 'https://university.harness.io/page/job-role-certifications'
export const HARNESS_COMMUNITY_LINK = 'https://community.harness.io/'
export const HARNESS_COMMUNITY_SLACK_LINK =
  'https://join.slack.com/t/harnesscommunity/shared_invite/zt-y4hdqh7p-RVuEQyIl5Hcx4Ck8VCvzBw'
export const HARNESS_DOCS_LINK = 'https://docs.harness.io/'
export const HARNESS_API_DOCS_LINK = 'https://apidocs.harness.io/'
export const HARNESS_TUTORIALS = 'https://developer.harness.io/tutorials/get-started'
export const SITE_STATUS_LINK = 'https://status.harness.io/'

export const timestamp = moment.now()
export const HARNESS_SUPPORT_LINK =
  '/sso.html?action=login&brand_id=114095000394&locale_id=1&return_to=https%3A%2F%2Fsupport.harness.io%2Fhc%2Fen-us%2Frequests&src=zendesk&timestamp=' +
  timestamp

export const WHATS_NEW_LINK = `https://docs.harness.io/article/ueeiji09do-what-s-new`
export const EARLY_ACCESS_LINK = `https://docs.harness.io/article/w4krvu96i3-early-access`

export const openZendeskSupport = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(HARNESS_SUPPORT_LINK)
}

export function getReleaseNodeLink(): string {
  switch (window.deploymentType) {
    case 'COMMUNITY':
    case 'ON_PREM': {
      return ON_PREM_RELEASE_NODE_LINK
    }
    default:
      return SAAS_RELEASE_NODE_LINK
  }
}
export const openFileATicket = (
  e: React.MouseEvent<Element, MouseEvent>,
  currentUserInfo: UserInfo,
  setShow?: React.Dispatch<React.SetStateAction<boolean>>
): void => {
  e.stopPropagation()
  e.preventDefault()
  window.Saber.do?.('set_options', {
    feedback_values: {
      Email: currentUserInfo.email // set default value for email field
    }
  })
  setShow?.(false)
  window.Saber.do?.('open')
}
export const openWhatsNew = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(WHATS_NEW_LINK)
}
export const openEarlyAccess = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(EARLY_ACCESS_LINK)
}
