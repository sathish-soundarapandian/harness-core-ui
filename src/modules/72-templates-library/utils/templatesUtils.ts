/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo, isEmpty } from 'lodash-es'
import type { UseStringsReturn } from 'framework/strings'
import type { TemplateSummaryResponse } from 'services/template-ng'
import templateFactory from '@templates-library/components/Templates/TemplatesFactory'
import type { Scope } from '@common/interfaces/SecretsInterface'

export enum TemplateType {
  Step = 'Step',
  Stage = 'Stage',
  Pipeline = 'Pipeline',
  CustomDeployment = 'CustomDeployment',
  MonitoredService = 'MonitoredService',
  SecretManager = 'SecretManager',
  Service = 'Service',
  Infrastructure = 'Infrastructure',
  StepGroup = 'StepGroup',
  Execution = 'Execution'
}

export enum TemplateUsage {
  USE = 'Use',
  COPY = 'Copy'
}

interface AllowedTemplate {
  label: string
  value: string
  disabled: boolean
}

export const getAllowedTemplateTypes = (
  scope: Scope,
  featureFlagBasedTemplates?: { [key: string]: boolean }
): { label: string; value: string; disabled?: boolean }[] => {
  const allowedTemplateTypes: AllowedTemplate[] = []
  Object.keys(TemplateType).forEach(item => {
    const template = templateFactory.getTemplate(item)
    const allowedScopes = template?.getAllowedScopes()
    if (allowedScopes && allowedScopes.includes(scope)) {
      allowedTemplateTypes.push({
        label: defaultTo(template?.getLabel(), ''),
        value: item,
        disabled: !template?.getIsEnabled() || featureFlagBasedTemplates?.[item] === false
      })
    }
  })
  return allowedTemplateTypes
}

export const getVersionLabelText = (
  template: TemplateSummaryResponse,
  getString: UseStringsReturn['getString']
): string | undefined => {
  return isEmpty(template.versionLabel)
    ? getString('templatesLibrary.alwaysUseStableVersion')
    : template.stableTemplate
    ? getString('templatesLibrary.stableVersion', { entity: template.versionLabel })
    : template.versionLabel
}

export const getTemplateRuntimeInputsCount = (templateInfo: { [key: string]: any }): number =>
  (JSON.stringify(templateInfo).match(/<\+input>/g) || []).length
