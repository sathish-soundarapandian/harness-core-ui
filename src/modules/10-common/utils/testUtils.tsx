/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { act, fireEvent, getByText, queryByAttribute, waitFor } from '@testing-library/react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation, useHistory } from 'react-router-dom'
import { ModalProvider } from '@harness/use-modal'
import qs from 'qs'
import { noop } from 'lodash-es'
import { enableMapSet } from 'immer'
import userEvent from '@testing-library/user-event'
import { AppStoreContext, AppStoreContextProps } from 'framework/AppStore/AppStoreContext'
import { LicenseStoreContext, LicenseStoreContextProps } from 'framework/LicenseStore/LicenseStoreContext'
import { LICENSE_STATE_VALUES } from 'framework/LicenseStore/licenseStoreUtil'
import { withAccountId, accountPathProps } from '@common/utils/routeUtils'
import type { Project } from 'services/cd-ng'
import { StringsContext, UseStringsReturn } from 'framework/strings'

import { FeaturesContext, FeaturesContextProps } from 'framework/featureStore/FeaturesContext'
import type { FeatureDetail, FeatureMetaData } from 'framework/featureStore/featureStoreUtil'
import type { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import './testUtils.scss'
import { PermissionsContext, PermissionsContextProps } from 'framework/rbac/PermissionsContext'
import { Editions } from '@common/constants/SubscriptionTypes'
import type { FeatureFlag } from '@common/featureFlags'
import { PreferenceStoreContext } from 'framework/PreferenceStore/PreferenceStoreContext'
import {
  TemplateSelectorContext,
  TemplatesSelectorContextInterface
} from 'framework/Templates/TemplateSelectorContext/TemplateSelectorContext'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseGetMockDataWithMutateAndRefetch<T> extends UseGetMockData<T> {
  mutate: () => Record<string, unknown>
  refetch: () => Record<string, unknown>
  cancel?: () => Record<string, unknown>
}

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): HTMLElement | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): HTMLElement | null => document.querySelector('.bp3-popover-content')

export interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, unknown>
  defaultAppStoreValues?: Partial<AppStoreContextProps>
  defaultLicenseStoreValues?: Partial<LicenseStoreContextProps>
  defaultPermissionValues?: Partial<PermissionsContextProps>
  defaultTemplateSelectorValues?: Partial<TemplatesSelectorContextInterface>
  defaultFeaturesValues?: Partial<FeaturesContextProps>
  defaultFeatureFlagValues?: Partial<Record<FeatureFlag, boolean>>
  projects?: Project[]
  enableBrowserView?: boolean
  stringsData?: Record<string, string>
  getString?: UseStringsReturn['getString']
}

export const prependAccountPath = (path: string): string => withAccountId(() => path)(accountPathProps)

export const CurrentLocation = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export interface BrowserViewProps {
  enable?: boolean
  children: React.ReactNode
}

export function BrowserView(props: BrowserViewProps): React.ReactElement {
  const { enable, children } = props
  const location = useLocation()
  const history = useHistory()

  if (!enable) {
    return <>{children}</>
  }

  function handlePathChange(e: React.ChangeEvent<HTMLInputElement>): void {
    history.replace(e.currentTarget.value)
  }

  const search = location.search ? `?${location.search.replace(/^\?/, '')}` : ''

  return (
    <div className="browser">
      <div className="browser-header">
        <input className="browser-path" value={location.pathname + search} onChange={handlePathChange} />
      </div>
      <div className="browser-content">{children}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  enableMapSet()
  const {
    path = '/',
    pathParams = {},
    defaultAppStoreValues,
    queryParams = {},
    defaultLicenseStoreValues,
    defaultPermissionValues,
    defaultTemplateSelectorValues,
    defaultFeaturesValues,
    defaultFeatureFlagValues = {},
    stringsData = {},
    getString = (key: string) => key
  } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const history = React.useMemo(() => createMemoryHistory({ initialEntries: [routePath] }), [])

  const defaultReturn = {
    enabled: true
  }

  /** TODO: Try fixing this later. This is causing some tests to fail */
  // React.useEffect(() => {
  //   history.replace(compile(path)(pathParams) + qs.stringify(queryParams, { addQueryPrefix: true }))
  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  // }, [path, pathParams, queryParams])

  return (
    <Router history={history}>
      <StringsContext.Provider value={{ data: stringsData as any, getString }}>
        <PreferenceStoreContext.Provider
          value={{
            set: noop,
            get: noop,
            clear: noop
          }}
        >
          <AppStoreContext.Provider
            value={{
              featureFlags: {
                FEATURE_ENFORCEMENT_ENABLED: true,
                ...defaultFeatureFlagValues
              },
              updateAppStore: () => void 0,
              currentUserInfo: { uuid: '' },
              ...defaultAppStoreValues
            }}
          >
            <LicenseStoreContext.Provider
              value={{
                versionMap: {},
                licenseInformation: {},
                CI_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                FF_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CCM_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CD_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CHAOS_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                STO_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CV_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CET_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                updateLicenseStore: () => void 0,
                ...defaultLicenseStoreValues
              }}
            >
              <PermissionsContext.Provider
                value={{
                  permissions: new Map<string, boolean>(),
                  requestPermission: () => void 0,
                  checkPermission: () => true,
                  cancelRequest: () => void 0,
                  ...defaultPermissionValues
                }}
              >
                <FeaturesContext.Provider
                  value={{
                    features: new Map<FeatureIdentifier, FeatureDetail>(),
                    featureMap: new Map<FeatureIdentifier, FeatureMetaData>(),
                    getEdition: () => {
                      return undefined
                    },
                    requestFeatures: () => void 0,
                    requestLimitFeature: () => void 0,
                    checkFeature: () => {
                      return defaultReturn
                    },
                    checkLimitFeature: () => {
                      return defaultReturn
                    },
                    getRestrictionType: () => {
                      return undefined
                    },
                    getHighestEdition: () => {
                      return Editions.FREE
                    },
                    ...defaultFeaturesValues
                  }}
                >
                  <ModalProvider>
                    <TemplateSelectorContext.Provider
                      value={{
                        state: {
                          isDrawerOpened: false,
                          selectorData: {
                            templateType: 'Step',
                            onSubmit: jest.fn(),
                            onCancel: jest.fn()
                          }
                        },
                        openTemplateSelector: jest.fn(),
                        closeTemplateSelector: jest.fn(),
                        ...defaultTemplateSelectorValues
                      }}
                    >
                      <RestfulProvider base="/">
                        <BrowserView enable={props.enableBrowserView}>
                          <Switch>
                            <Route exact path={path}>
                              {props.children}
                            </Route>
                            <Route>
                              <CurrentLocation />
                            </Route>
                          </Switch>
                        </BrowserView>
                      </RestfulProvider>
                    </TemplateSelectorContext.Provider>
                  </ModalProvider>
                </FeaturesContext.Provider>
              </PermissionsContext.Provider>
            </LicenseStoreContext.Provider>
          </AppStoreContext.Provider>
        </PreferenceStoreContext.Provider>
      </StringsContext.Provider>
    </Router>
  )
}

export const queryByNameAttribute = (name: string, container: HTMLElement): HTMLElement | null =>
  queryByAttribute('name', container, name)

export const doConfigureOptionsTesting = async (
  cogModal: HTMLElement,
  fieldElement: HTMLInputElement
): Promise<void> => {
  // Type regex and submit
  // check if field has desired value
  await waitFor(() => expect(getByText(cogModal, 'common.configureOptions.regex')).toBeInTheDocument())
  const regexRadio = getByText(cogModal, 'common.configureOptions.regex')
  userEvent.click(regexRadio)
  const regexTextArea = queryByAttribute('name', cogModal, 'regExValues') as HTMLInputElement
  act(() => {
    fireEvent.change(regexTextArea, { target: { value: '<+input>.includes(/test/)' } })
  })
  const cogSubmit = getByText(cogModal, 'submit')
  userEvent.click(cogSubmit)
  await waitFor(() => expect(fieldElement.value).toBe('<+input>.regex(<+input>.includes(/test/))'))
}
