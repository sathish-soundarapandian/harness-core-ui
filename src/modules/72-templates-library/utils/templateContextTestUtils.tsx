/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { Route } from 'react-router-dom'

import { pick } from 'lodash-es'

import { VisualYamlSelectedView } from '@harness/uicore'
import type { TestWrapperProps } from '@common/utils/testUtils';
import { TestWrapper } from '@common/utils/testUtils'
import type {
  TemplateContextInterface
} from '@templates-library/components/TemplateStudio/TemplateContext/TemplateContext';
import {
  TemplateContext
} from '@templates-library/components/TemplateStudio/TemplateContext/TemplateContext'

import templateContextProviderProps from './__tests__/mocks/templateContextProvider.json'

export interface TemplateContextWrapperProps extends TestWrapperProps {
  templateContextValues?: Partial<TemplateContextInterface>
  isGitSyncEnabled?: boolean
}

export const TemplateContextTestWrapper: React.FC<TemplateContextWrapperProps> = props => {
  const { defaultAppStoreValues, templateContextValues, isGitSyncEnabled = true } = props
  const [view, setView] = React.useState<VisualYamlSelectedView>(VisualYamlSelectedView.VISUAL)
  return (
    <TestWrapper
      {...pick(props, ['path', 'pathParams', 'queryParams'])}
      defaultAppStoreValues={{
        featureFlags: {},
        selectedProject: {
          identifier: 'dummy',
          name: 'dummy',
          modules: ['CD']
        },
        isGitSyncEnabled: isGitSyncEnabled,
        connectivityMode: 'DELEGATE',
        ...defaultAppStoreValues
      }}
    >
      <TemplateContext.Provider
        value={
          {
            view: view,
            setView: setView,
            updateTemplateView: jest.fn(),
            ...templateContextProviderProps,
            ...templateContextValues,
            state: {
              ...templateContextProviderProps.state,
              ...templateContextValues?.state,
              yamlHandler: {
                getLatestYaml: () => 'dummy\n',
                getYAMLValidationErrorMap: () => new Map(),
                ...templateContextValues?.state?.yamlHandler
              }
            }
          } as TemplateContextInterface
        }
      >
        <Route exact path={props.path}>
          {props.children}
        </Route>
      </TemplateContext.Provider>
    </TestWrapper>
  )
}
