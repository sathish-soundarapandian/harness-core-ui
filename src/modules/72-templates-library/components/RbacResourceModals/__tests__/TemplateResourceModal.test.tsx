/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, waitFor } from '@testing-library/react'
import { SortMethod } from '@harness/uicore'
import { TestWrapper } from '@common/utils/testUtils'
import { defaultAppStoreValues } from '@common/utils/DefaultAppStoreData'
import TemplateResourceModal from '../TemplateResourceModal'
import mockData from './templateResourceMockData.json'

const props = {
  searchTerm: '',
  sortMethod: SortMethod.Newest,
  onSelectChange: jest.fn(),
  selectedData: [],
  resourceScope: {
    accountIdentifier: ''
  }
}

const params = {
  accountId: 'testAcc',
  orgIdentifier: 'testOrg',
  projectIdentifier: 'test',
  module: 'cd'
}
const mockGetCallFunction = jest.fn()
jest.useFakeTimers()
jest.mock('services/template-ng', () => ({
  useGetTemplateList: jest.fn().mockImplementation(args => {
    mockGetCallFunction(args)
    return { mutate: jest.fn(() => Promise.resolve(mockData)), cancel: jest.fn(), loading: false }
  })
}))

describe('<TemplateResourceModal /> Tests', () => {
  test('snapshot test', async () => {
    const { getByText, container } = render(
      <TestWrapper pathParams={params} defaultAppStoreValues={defaultAppStoreValues}>
        <TemplateResourceModal {...props} />
      </TestWrapper>
    )
    await waitFor(() => getByText('common.templates'))
    jest.runOnlyPendingTimers()

    expect(container).toMatchSnapshot()
  })
})
