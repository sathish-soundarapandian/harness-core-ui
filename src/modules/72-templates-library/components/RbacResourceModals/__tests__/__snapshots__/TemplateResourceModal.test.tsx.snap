// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateResourceModal /> Tests snapshot test 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main"
  >
    <div
      class="TableV2--table"
    >
      <div
        class="TableV2--header"
        role="row"
      >
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 5%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            Â 
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 40%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            common.templates
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 30%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            common.entityType
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 20%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            templatesLibrary.templateScope
          </p>
        </div>
      </div>
      <div
        class="TableV2--body"
      >
        <div
          class="TableV2--row TableV2--card TableV2--clickable"
          role="row"
        >
          <div
            class="TableV2--cells"
          >
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 5%;"
            >
              <label
                class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox checkBox StyledProps--main"
              >
                <input
                  type="checkbox"
                />
                <span
                  class="bp3-control-indicator"
                />
              </label>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 40%;"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main"
                data-testid="CI_Stage_Template"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
                >
                  <span
                    class="bp3-popover-wrapper"
                  >
                    <span
                      class="bp3-popover-target Text--targetClass"
                    >
                      <p
                        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                        tabindex="0"
                      >
                        CI Stage Template 2
                      </p>
                    </span>
                  </span>
                </div>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400"
                >
                  idLabel
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 30%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  Stage
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 20%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  project
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="TableV2--row TableV2--card TableV2--clickable"
          role="row"
        >
          <div
            class="TableV2--cells"
          >
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 5%;"
            >
              <label
                class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox checkBox StyledProps--main"
              >
                <input
                  type="checkbox"
                />
                <span
                  class="bp3-control-indicator"
                />
              </label>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 40%;"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main"
                data-testid="new_step_name"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
                >
                  <span
                    class="bp3-popover-wrapper"
                  >
                    <span
                      class="bp3-popover-target Text--targetClass"
                    >
                      <p
                        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                        tabindex="0"
                      >
                        New Step Name
                      </p>
                    </span>
                  </span>
                </div>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400"
                >
                  idLabel
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 30%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  Step
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 20%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  project
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="TableV2--row TableV2--card TableV2--clickable"
          role="row"
        >
          <div
            class="TableV2--cells"
          >
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 5%;"
            >
              <label
                class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox checkBox StyledProps--main"
              >
                <input
                  type="checkbox"
                />
                <span
                  class="bp3-control-indicator"
                />
              </label>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 40%;"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main"
                data-testid="CI_Stage"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
                >
                  <span
                    class="bp3-popover-wrapper"
                  >
                    <span
                      class="bp3-popover-target Text--targetClass"
                    >
                      <p
                        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                        tabindex="0"
                      >
                        CI Stage
                      </p>
                    </span>
                  </span>
                </div>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400"
                >
                  idLabel
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 30%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  Stage
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 20%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  project
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="TableV2--row TableV2--card TableV2--clickable"
          role="row"
        >
          <div
            class="TableV2--cells"
          >
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 5%;"
            >
              <label
                class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox checkBox StyledProps--main"
              >
                <input
                  type="checkbox"
                />
                <span
                  class="bp3-control-indicator"
                />
              </label>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 40%;"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main"
                data-testid="CD_Stage"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
                >
                  <span
                    class="bp3-popover-wrapper"
                  >
                    <span
                      class="bp3-popover-target Text--targetClass"
                    >
                      <p
                        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                        tabindex="0"
                      >
                        CD Stage
                      </p>
                    </span>
                  </span>
                </div>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400"
                >
                  idLabel
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 30%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  Stage
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 20%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  project
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="TableV2--row TableV2--card TableV2--clickable"
          role="row"
        >
          <div
            class="TableV2--cells"
          >
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 5%;"
            >
              <label
                class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox checkBox StyledProps--main"
              >
                <input
                  type="checkbox"
                />
                <span
                  class="bp3-control-indicator"
                />
              </label>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 40%;"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main"
                data-testid="templatetest2"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
                >
                  <span
                    class="bp3-popover-wrapper"
                  >
                    <span
                      class="bp3-popover-target Text--targetClass"
                    >
                      <p
                        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                        tabindex="0"
                      >
                        template-test2
                      </p>
                    </span>
                  </span>
                </div>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400"
                >
                  idLabel
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 30%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  Step
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 20%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  project
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="TableV2--row TableV2--card TableV2--clickable"
          role="row"
        >
          <div
            class="TableV2--cells"
          >
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 5%;"
            >
              <label
                class="bp3-control bp3-checkbox StyledProps--font Checkbox--checkbox checkBox StyledProps--main"
              >
                <input
                  type="checkbox"
                />
                <span
                  class="bp3-control-indicator"
                />
              </label>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 40%;"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main"
                data-testid="Echo_Template"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
                >
                  <span
                    class="bp3-popover-wrapper"
                  >
                    <span
                      class="bp3-popover-target Text--targetClass"
                    >
                      <p
                        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                        tabindex="0"
                      >
                        Echo Template
                      </p>
                    </span>
                  </span>
                </div>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400"
                >
                  idLabel
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 30%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  Step
                </p>
              </div>
            </div>
            <div
              class="TableV2--cell"
              role="cell"
              style="width: 20%;"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey800"
                >
                  project
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
